src/Common/oclBasics.cpp

```diff
@@ -0,0 +1,378 @@
+            CASE_CL_CONSTANT(CL_INVALID_EVENT_WAIT_LIST)
+            CASE_CL_CONSTANT(CL_INVALID_EVENT)
+            CASE_CL_CONSTANT(CL_INVALID_OPERATION)
+            CASE_CL_CONSTANT(CL_INVALID_GL_OBJECT)
+            CASE_CL_CONSTANT(CL_INVALID_BUFFER_SIZE)
+            CASE_CL_CONSTANT(CL_INVALID_MIP_LEVEL)
+            CASE_CL_CONSTANT(CL_INVALID_GLOBAL_WORK_SIZE)
+            CASE_CL_CONSTANT(CL_INVALID_PROPERTY)
+            CASE_CL_CONSTANT(CL_INVALID_IMAGE_DESCRIPTOR)
+            CASE_CL_CONSTANT(CL_INVALID_COMPILER_OPTIONS)
+            CASE_CL_CONSTANT(CL_INVALID_LINKER_OPTIONS)
+            CASE_CL_CONSTANT(CL_INVALID_DEVICE_PARTITION_COUNT)
+            default:
+                return "UNKNOWN ERROR CODE ";
+        }
+
+#undef CASE_CL_CONSTANT
+    }
+
+
+    static void checkError(cl_int error)
+    {
+        if (error != CL_SUCCESS)
+            throw DB::Exception("OpenCL error " + opencl_error_to_str(error), DB::ErrorCodes::OPENCL_ERROR);
+    }
+
+
+    /// Getting OpenCl main entities.
+
+    static cl_platform_id getPlatformID(const Settings & settings)
+    {
+        cl_platform_id platform;
+        cl_int error = clGetPlatformIDs(settings.numberOfPlatformEntries, &platform,
+                                        settings.numberOfAvailablePlatforms);
+        checkError(error);
+
+        return platform;
+    }
+
+
+    static cl_device_id getDeviceID(cl_platform_id & platform, const Settings & settings)
+    {
+        cl_device_id device;
+        cl_int error = clGetDeviceIDs(platform, CL_DEVICE_TYPE_GPU, settings.numberOfDevicesEntries,
+                                      &device, settings.numberOfAvailableDevices);
+        OCL::checkError(error);
+
+        return device;
+    }
+
+
+    static cl_context makeContext(cl_device_id & device, const Settings & settings)
+    {
+        cl_int error;
+        cl_context gpuContext = clCreateContext(settings.contextProperties, settings.numberOfDevicesEntries,
+                                                &device, settings.contextCallback, settings.contextCallbackData,
+                                                &error);
+        OCL::checkError(error);
+
+        return gpuContext;
+    }
+
+
+    static cl_command_queue makeCommandQueue(cl_device_id & device, cl_context & context, const Settings & settings)
+    {
+        cl_int error;
+#ifdef CL_USE_DEPRECATED_OPENCL_1_2_APIS
+        cl_command_queue commandQueue = clCreateCommandQueue(context, device, settings.commandQueueProperties, &error);
+#else
+        cl_command_queue commandQueue = clCreateCommandQueueWithProperties(context, device, nullptr, &error);
+#endif
+        OCL::checkError(error);
+
+        return commandQueue;
+    }
+
+
+    static cl_program makeProgram(const char * file_name, cl_context context,
+                                  cl_device_id device_id, const Settings & settings)
+    {
+        FILE * file = fopen(file_name, "rb");
```

Comment:
This logic does not work. We should not use resources. Define kernels as constant strings.